### DISCUSSION TITLE
Dynamic DAG creation for Schedule task on DB

### DISCUSSION THREAD

**Discussion by jcastagomez** [apache/airflow#25574] [Q&A]
Hi Guys I’m new working on Airflow. I’m looking for a way to schedule jobs stored on a DB with their corresponding cron expression. So I was looking for a create method on the AIRFLOW REST API, but It seems that It’s not possible. So My understanding is that It’s only possible by using Dynamc DAG Creation (with globals/single file creation, env variables, files, etc). Digging into that, I noticed that It’s not a good practice to use globals or the single file creation because Airflow parses every n seconds all the files on the folder, so It could generate performance issues. For that reason, I’m thinking about using multiple file creation, getting the DAGS  configuration from the DB and then create a pipeline to publish the created DAGS on Airflow, but I’m not pretty sure if that is the best approach having into account Airflow best practices.

I would really appreciate all your suggestion about it.

**Comment by potiuk** [comment]
I think all options that make sense are nicely described in https://airflow.apache.org/docs/apache-airflow/stable/howto/dynamic-dag-generation.html?highlight=dynamic+dags -> pick one that fits you best. Generally avoid any DAG top-level code that is synchronously reaching out to a database or smth like that, put all the data needed to generate your dynamic DAGs in folder (either as part of generated python code or some structured (json files?) metadata.

This will work best in various DAG synchronisation scheme.

Also if you are looking for some optimisations and reducing worker-overhead on a single DAG file generating multiple dags you can read those two articles:

* https://medium.com/apache-airflow/airflows-magic-loop-ec424b05b629
* https://medium.com/apache-airflow/magic-loop-in-airflow-reloaded-3e1bd8fb6671

And for Airlfow 2.4 we will have an experimental support to make it a bit more easier and more "predictable" by the Airflow code  https://github.com/apache/airflow/pull/25161 - particularly this part of the updated documentation might be interesting for you https://github.com/apache/airflow/pull/25161/files#diff-fcdb96b2d394b0a5412a959209172aa42848ead34f3bbbfb2802e953366bb3f6R143 (but this "util" will be available only in Airlfow 2.4 and subject to removal/update if we implement some better optimisations).

**Reply by jcastagomez** [reply]
thanks @potiuk for your prompt response , that was the doc that I was reading about. My main concern about creating DAGs by using  the single method is that I need to get all my schedule objects in the Top of the DAG from the source DB or REST API. So the DB or rest API will be call n times. The another important think to take into account is that the source DB or rest API is changing all the time (scheduled objects can be changed, deleted or new object can be added so There's an important challenge about the synchronization as well.


schedule_objects = read_schedule_objects_from_db_or_rest_api()

start_time = default_timer()
for schedule_id, name, cron_exp in schedule_objects:
    dag_id = f"schedule_{schedule_id}"
    dag = DAG(
      dag_id,
      tags=[name],
      ...
    )
    with dag:
        task_1 = PythonOperator(...)
        tesk_2 = BashOperator(...)
        task_3 = SingleHttpOperator(...)
        # more tasks
        ...
        task_1 >> [task_2, task_3]
    globals()[dag_id] = dag
took = default_timer() - start_time
logger.info(f"took: {took}")

**Reply by potiuk** [reply]
Read the docs carefully. You misunderstood it.. you have to revert your thinking and periodicaly dump result of your DB/REST API call as file into the DAG folder not have the DAG poll the data.

**Reply by jcastagomez** [reply]
Thanks @potiuk  I'll dig into that.
